nixos/modules/services/security/crowdsec.nix

```diff
@@ -0,0 +1,959 @@
+                };
+              };
+            }) localParsersS00RawMap
+          )
+          // builtins.listToAttrs (
+            map (parser: {
+              inherit cfg;
+              name = lib.strings.normalizePath "${localParsersS01ParseDir}/${builtins.unsafeDiscardStringContext (builtins.baseNameOf parser)}";
+              value = {
+                link = {
+                  type = "L+";
+                  argument = "${parser}";
+                };
+              };
+            }) localParsersS01ParseMap
+          )
+          // builtins.listToAttrs (
+            map (parser: {
+              inherit cfg;
+              name = lib.strings.normalizePath "${localParsersS02EnrichDir}/${builtins.unsafeDiscardStringContext (builtins.baseNameOf parser)}";
+              value = {
+                link = {
+                  type = "L+";
+                  argument = "${parser}";
+                };
+              };
+            }) localParsersS02EnrichMap
+          )
+          // builtins.listToAttrs (
+            map (postoverflow: {
+              inherit cfg;
+              name = lib.strings.normalizePath "${localPostOverflowsS01WhitelistDir}/${builtins.unsafeDiscardStringContext (builtins.baseNameOf postoverflow)}";
+              value = {
+                link = {
+                  type = "L+";
+                  argument = "${postoverflow}";
+                };
+              };
+            }) localPostOverflowsS01WhitelistMap
+          )
+          // builtins.listToAttrs (
+            map (context: {
+              inherit cfg;
+              name = lib.strings.normalizePath "${localContextsDir}/${builtins.unsafeDiscardStringContext (builtins.baseNameOf context)}";
+              value = {
+                link = {
+                  type = "L+";
+                  argument = "${context}";
+                };
+              };
+            }) localContextsMap
+          )
+          // builtins.listToAttrs (
+            map (notification: {
+              inherit cfg;
+              name = lib.strings.normalizePath "${notificationsDir}/${builtins.unsafeDiscardStringContext (builtins.baseNameOf notification)}";
+              value = {
+                link = {
+                  type = "L+";
+                  argument = "${notification}";
+                };
+              };
+            }) localNotificationsMap
+          );
+      };
+
+      users.users.${cfg.user} = {
+        name = cfg.user;
+        description = lib.mkDefault "CrowdSec service user";
+        isSystemUser = true;
+        group = cfg.group;
+        extraGroups = [ "systemd-journal" ];
+      };
+
+      users.groups.${cfg.group} = lib.mapAttrs (name: lib.mkDefault) { };
+
+      networking.firewall.allowedTCPPorts =
+        let
+          parsePortFromURLOption =
+            url: option:
+            builtins.addErrorContext "`${option}` requires a port to be specified, but we failed to parse a port from '${url}'" (
```

Comment:
`addErrorContext` text should usually be phrased as a "while ...", e.g. "while extracting a port from URL: ${url}".

grepping treewide for `addErrorContext` may yield some inspiration and/or best practices.

---

Q1: where does url come from? Would it make sense for "has a port suffix" to be part of the relevant option-type's `check` function?

Q2: do we need to handle URLs that have additional stuff _after_ the port segment? E.g. `/`paths, `?`queries, `#`anchors, etc. _I guess_ `toInt` will handle those by throwing an error... Is that acceptable here?

Reply:
> The one that @MattSturgeon was referring to. I'm guessing he meant the check function for the URL option, but we don't have a URL option, just a general settingsoption...

I'm sorry but I'm still not really able to follow you. Do you mean with
> I'm guessing he meant the check function for the URL option

the `parsePortFromURLOption` function?

Reply:
Ooh, ok, that makes more sense now.

> So I guess we just need to add an assert with a nice message to the `parsePortFromURLOption` function.

I'm so sorry but... I think that I'm not understanding your intention here :sob: .
In my eyes, the url is already correctly checked and is wrapped in a context with a custom error message.

Reply:
Then let's ignore it. I wasn't the one who suggested it anyway 😅.